### PR TITLE
CF-2471: Added pre validation

### PR DESCRIPTION
**Pull request by ihor-yamshchykov**
Added onValidationFailed function;
Added required validator;
Prevented validation of non-required fields if they are not dirty;
Added validation on form init;

### CODE REVIEW

File: dist/lib/validation.js [hunk 1]
```diff
@@ -118,7 +125,18 @@ var getActiveQuestionsFromQuestionSets = function getActiveQuestionsFromQuestion
     return Array.prototype.push.apply(questionsToCheck, getActiveQuestions(questionSet.questions, questionAnswers));
   });
 
-  return questionsToCheck;
+  return questionsToCheck.map(function (question) {
+    return Object.assign({}, question, { answer: questionAnswers[question.questionId] });
```
Comment (bdryl-softserve) [inline dist/lib/validation.js]: Spread?

File: dist/lib/validation.js [hunk 2]
```diff
@@ -118,7 +125,18 @@ var getActiveQuestionsFromQuestionSets = function getActiveQuestionsFromQuestion
...
+  });
+};
+
+/**
+ * Will validate questions that are required and not empty and have non-empty validators
+ */
+var shouldValidateQuestion = function shouldValidateQuestion(question) {
```
Comment (bdryl-softserve) [inline dist/lib/validation.js]: const?

File: dist/lib/validation.js [hunk 3]
```diff
@@ -118,7 +125,18 @@ var getActiveQuestionsFromQuestionSets = function getActiveQuestionsFromQuestion
...
+  return question.validations instanceof Array && question.validations.length > 0 && (question.validations.some(function (val) {
```
Comment (bdryl-softserve) [inline dist/lib/validation.js]: Let's split this line to smaller parts.

File: dist/lib/validation.js (hunk 1 above)
Comment (ihor-yamshchykov) [inline dist/lib/validation.js]: dist/lib/validation.js
and there's no preset

File: dist/lib/validation.js (hunk 2 above)
Comment (ihor-yamshchykov) [inline dist/lib/validation.js]: dist/lib/validation.js

**Comment by macku** [comment]
Moved to https://github.com/mudynamics/WinterfellCF/pull/3

**Pull request closed without merging**